enhancements/workload-partitioning/rationalizing-configuration.md

```diff
@@ -0,0 +1,478 @@
+consideration in environments with short, fixed maintenance
+windows). To continue to support enabling the feature during
+deployment, and to simplify the configuration, we propose to extend
+the PAO with a `render` command, similar to the one in other
+operators, to have it generate manifests for the OpenShift installer
+to consume. This avoids any need to change the OpenShift installer to
+support this uncommon use case.
+
+The `render` command will have an option to enable the management
+partition. When the option is given, the `render` command will create
+a manifest for the new enablement API with status data that includes
+the `management` workload type to enable partitioning (the details of
+that API will be defined in a separate design document). In the
+future, the PAO might enable partitioning for all of the types
+mentioned in the input PerformanceProfiles, or based on some other
+input.
+
+For each PerformanceProfile, the PAO will render MachineConfig
+manifests with higher precedence than the default ones created by the
+installer during bootstrapping. These manifests will include all of
+the details for configuring `kubelet`, CRI-O, and `tuned` to know
+about and partition workloads, including the cpusets for each workload
+type.
+
+The PAO `render` command also generates a manifest with the CRD for
+PerformanceProfile so bootstrapping does not block when trying to
+install the PerformanceProfile manifests.
+
+### Future Work
+
+1. Write an enhancement to describe the API the admission hook will
+   use to determine when workload partitioning is enabled.
+2. Write an enhancement to describe the changes in
+   performance-addon-operator to manage the kubelet and CRI-O
+   configuration when workload partitioning is enabled, including the
+   `render` command. See https://issues.redhat.com/browse/CNF-2164 to
+   track that work.
+
+### User Stories
+
+#### Enabling and Configuring at the Same Time
+
+In this workflow, the user provides all of the information needed to
+fully partition the workloads from the very beginning.
+
+1. User runs the installer to `create manifests` to get standard
+   manifests without any partitioning.
+2. The user creates PerformanceProfile manifests for each known
+   MachineConfigPool and adds them to the installer inputs.
+3. The user runs the PAO `render` command to read the
+   PerformanceProfile manifests and create additional manifests, as
+   described above.
+4. User runs the installer to `create cluster`.
+5. Installer bootstraps the cluster.
+6. Bootstrapping runs the machine-config-operator (MCO) `render`
+   command, which generates MachineConfig manifests with low
+   precedence.
+7. Bootstrapping uploads both sets of MachineConfig manifests, one
+   after the other.
+8. The machine-config-operator (MCO) applies MachineConfigs to nodes
+   in the cluster.
+9. Bootstrapping finishes and the cluster is launched.
+10. If the MCO does not apply KubeletConfig in step 8, it must do it
```

Comment:
I'm a bit confused. Up until now you do not explicitly mention kubeletconfigs, but instead use machineconfigs. Is this what the PAO is generating?

If so this is today not doable as a day 1 operation, and will effectively be day2 (extra reboots across all nodes) but will be automatic

Reply:
We did have some chatter about allowing day-1 kubeletconfig generation (adding the kubeletconfigcontroller into the bootstrap MCC), so that's a potential option if really needed

Reply:
> We did have some chatter about allowing day-1 kubeletconfig generation (adding the kubeletconfigcontroller into the bootstrap MCC), so that's a potential option if really needed

Yes, we're trying to simultaneously configure things properly from the very beginning of the cluster's life (to avoid some very ugly edge cases) and avoid a ton of reboots (to minimize deploy time, since metal reboots veeeery slowly).